File: app/src/main/java/ru/SnowVolf/convertx/ui/fragments/base/BaseMainFragment.java
package ru.SnowVolf.convertx.ui.fragments.base;

import android.os.Bundle;
import android.support.design.widget.TextInputEditText;
import android.support.v7.view.menu.MenuBuilder;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Toast;

import ru.SnowVolf.convertx.R;
import ru.SnowVolf.convertx.settings.Preferences;
import ru.SnowVolf.convertx.ui.Toasty;
import ru.SnowVolf.convertx.utils.StringUtils;

/**
 * Created by Snow Volf on 20.06.2017, 11:35
 */

public class BaseMainFragment extends BaseFragment {
    public TextInputEditText mInputField, mOutputField;
    public Button mCopyInput, mCopyOutput, mClearOutput, mPasteInput, mPasteOutput;
    public BaseMainFragment(){}


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState){
        rootView = inflater.inflate(R.layout.fragment_main, container, false);
        mInputField.setHint(R.string.hint_utf);
        mOutputField.setHint(R.string.hint_unicode);
        mInputField = (TextInputEditText) rootView.findViewById(R.id.inputText);
        mOutputField = (TextInputEditText) rootView.findViewById(R.id.outputText);
        mCopyInput = (Button) rootView.findViewById(R.id.copy_btn2);
        mCopyOutput = (Button) rootView.findViewById(R.id.copy_btn);
        mClearOutput = (Button) rootView.findViewById(R.id.clear_btn);
        mPasteInput = (Button) rootView.findViewById(R.id.paste_btn);
        mPasteOutput = (Button) rootView.findViewById(R.id.paste_btn2);
        return rootView;
    }

    @Override
    public void onSaveInstanceState(Bundle outState){
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        setHasOptionsMenu(true);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState){
        super.onActivityCreated(savedInstanceState);
        mCopyInput.setOnClickListener(this::copyInput);
        mCopyOutput.setOnClickListener(this::copyOutput);
        mClearOutput.setOnClickListener(view -> clearOutput());
        mPasteInput.setOnClickListener(view -> pasteInput());
        mPasteOutput.setOnClickListener(view -> pasteOutput());
        mOutputField.setTypeface(Mono);
        mOutputField.setTypeface(Mono);
        mOutputField.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                if (charSequence.toString().equals("")){
                    mPasteOutput.setVisibility(View.VISIBLE);
                    mClearOutput.setVisibility(View.INVISIBLE);
                } else {
                    mPasteOutput.setVisibility(View.INVISIBLE);
                    mClearOutput.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });
    }

    @Override
    public void onResume(){
        super.onResume();
        if (mInputField != null && mOutputField != null){
            try {
                mInputField.setTextSize(Preferences.getFontSize());
                mOutputField.setTextSize(Preferences.getFontSize());
            } catch (Exception ex){
                ex.printStackTrace();
            }
        }
    }

    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        if (menu != null) {
            menu.clear();
        } else
            menu = new MenuBuilder(getContext());
        //добавляем пункты меню
        menu.add(R.string.convert)
                .setIcon(R.drawable.ic_menu_convert)
                .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS)
                .setOnMenuItemClickListener(menuItem -> {
                    Convert();
                    return true;
                });
        menu.add(R.string.clear_all)
                .setIcon(R.drawable.ic_menu_clear_all)
                .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS)
                .setOnMenuItemClickListener(menuItem -> {
                    clearAllText();
                    return true;
                });
    }

    public void Convert() {
    }

    public void copyInput(View v){
        StringUtils.INSTANCE.copyToClipboard(getContext(), mInputField.getText().toString());
        Toasty.info(getContext(), getString(R.string.copied2clipboard), Toast.LENGTH_SHORT, true).show();
    }
    public void copyOutput(View v){
        StringUtils.INSTANCE.copyToClipboard(getContext(), mOutputField.getText().toString());
        Toasty.info(getContext(), getString(R.string.copied2clipboard), Toast.LENGTH_SHORT, true).show();
    }

    public void clearOutput(){
        mOutputField.setText("");
    }

    public void pasteInput(){
        try{ mInputField.setText(StringUtils.INSTANCE.readFromClipboard()); } catch (NullPointerException ignored){}
    }

    public void pasteOutput(){
        try{ mOutputField.setText(StringUtils.INSTANCE.readFromClipboard()); } catch (NullPointerException ignored){}
    }

    public void clearAllText(){
        mOutputField.setText("");
        mInputField.setText("");
        Toasty.success(getContext(), getString(R.string.cleared), Toast.LENGTH_SHORT, true).show();
    }
}

File: app/src/main/java/ru/SnowVolf/convertx/ui/fragments/base/BaseMainFragment.kt
package ru.SnowVolf.convertx.ui.fragments.base

import ru.SnowVolf.convertx.utils.StringUtils.copyToClipboard
import ru.SnowVolf.convertx.utils.StringUtils.readFromClipboard
import com.google.android.material.textfield.TextInputEditText
import android.os.Bundle
import ru.SnowVolf.convertx.R
import android.text.TextWatcher
import android.text.Editable
import android.view.*
import android.widget.Button
import ru.SnowVolf.convertx.ui.Toasty
import android.widget.Toast
import androidx.appcompat.view.menu.MenuBuilder
import ru.SnowVolf.convertx.settings.Preferences
import java.lang.Exception
import java.lang.NullPointerException

/**
 * Created by Snow Volf on 20.06.2017, 11:35
 */
open class BaseMainFragment : BaseFragment() {
    @JvmField
    var mInputField: TextInputEditText? = null
    @JvmField
    var mOutputField: TextInputEditText? = null
    @JvmField
    var mCopyInput: Button? = null
    @JvmField
    var mCopyOutput: Button? = null
    @JvmField
    var mClearOutput: Button? = null
    @JvmField
    var mPasteInput: Button? = null
    @JvmField
    var mPasteOutput: Button? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(R.layout.fragment_main, container, false)
        mInputField!!.setHint(R.string.hint_utf)
        mOutputField!!.setHint(R.string.hint_unicode)
        mInputField = rootView.findViewById<View>(R.id.inputText) as TextInputEditText
        mOutputField = rootView.findViewById<View>(R.id.outputText) as TextInputEditText
        mCopyInput = rootView.findViewById<View>(R.id.copy_btn2) as Button
        mCopyOutput = rootView.findViewById<View>(R.id.copy_btn) as Button
        mClearOutput = rootView.findViewById<View>(R.id.clear_btn) as Button
        mPasteInput = rootView.findViewById<View>(R.id.paste_btn) as Button
        mPasteOutput = rootView.findViewById<View>(R.id.paste_btn2) as Button
        return rootView
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
        setHasOptionsMenu(true)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        mCopyInput!!.setOnClickListener { v: View? -> copyInput(v) }
        mCopyOutput!!.setOnClickListener { v: View? -> copyOutput(v) }
        mClearOutput!!.setOnClickListener { view: View? -> clearOutput() }
        mPasteInput!!.setOnClickListener { view: View? -> pasteInput() }
        mPasteOutput!!.setOnClickListener { view: View? -> pasteOutput() }
        mOutputField!!.typeface = Mono
        mOutputField!!.typeface = Mono
        mOutputField!!.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                if (charSequence.toString() == "") {
                    mPasteOutput!!.visibility = View.VISIBLE
                    mClearOutput!!.visibility = View.INVISIBLE
                } else {
                    mPasteOutput!!.visibility = View.INVISIBLE
                    mClearOutput!!.visibility = View.VISIBLE
                }
            }

            override fun afterTextChanged(editable: Editable) {}
        })
    }

    override fun onResume() {
        super.onResume()
        if (mInputField != null && mOutputField != null) {
            try {
                mInputField!!.textSize = Preferences.fontSize.toFloat()
                mOutputField!!.textSize = Preferences.fontSize.toFloat()
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.clear()
        //добавляем пункты меню
        menu.add(R.string.convert)
            .setIcon(R.drawable.ic_menu_convert)
            .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS)
            .setOnMenuItemClickListener { menuItem: MenuItem? ->
                Convert()
                true
            }
        menu.add(R.string.clear_all)
            .setIcon(R.drawable.ic_menu_clear_all)
            .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS)
            .setOnMenuItemClickListener { menuItem: MenuItem? ->
                clearAllText()
                true
            }
    }

    open fun Convert() {}
    fun copyInput(v: View?) {
        copyToClipboard(requireContext(), mInputField!!.text.toString())
        Toasty.info(requireContext(), getString(R.string.copied2clipboard), Toast.LENGTH_SHORT, true).show()
    }

    fun copyOutput(v: View?) {
        copyToClipboard(requireContext(), mOutputField!!.text.toString())
        Toasty.info(requireContext(), getString(R.string.copied2clipboard), Toast.LENGTH_SHORT, true).show()
    }

    fun clearOutput() {
        mOutputField!!.setText("")
    }

    fun pasteInput() {
        try {
            mInputField!!.setText(readFromClipboard())
        } catch (ignored: NullPointerException) {
        }
    }

    fun pasteOutput() {
        try {
            mOutputField!!.setText(readFromClipboard())
        } catch (ignored: NullPointerException) {
        }
    }

    open fun clearAllText() {
        mOutputField!!.setText("")
        mInputField!!.setText("")
        Toasty.success(requireContext(), getString(R.string.cleared), Toast.LENGTH_SHORT, true).show()
    }
}